clients/src/test/java/org/apache/kafka/clients/NetworkClientApiVersionsCheckTest.java

```diff
@@ -0,0 +1,102 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.clients;
+
+
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.network.NetworkReceive;
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.apache.kafka.common.protocol.ProtoUtils;
+import org.apache.kafka.common.protocol.Protocol;
+import org.apache.kafka.common.protocol.types.Struct;
+import org.apache.kafka.common.requests.ApiVersionsResponse;
+import org.apache.kafka.common.requests.ResponseHeader;
+import org.apache.kafka.test.DelayedReceive;
+import org.junit.Test;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+import static org.apache.kafka.common.requests.ApiVersionsResponse.API_KEY_NAME;
+import static org.apache.kafka.common.requests.ApiVersionsResponse.API_VERSIONS_KEY_NAME;
+import static org.apache.kafka.common.requests.ApiVersionsResponse.ERROR_CODE_KEY_NAME;
+import static org.apache.kafka.common.requests.ApiVersionsResponse.MAX_VERSION_KEY_NAME;
+import static org.apache.kafka.common.requests.ApiVersionsResponse.MIN_VERSION_KEY_NAME;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class NetworkClientApiVersionsCheckTest extends NetworkClientTest {
+
+    private static final ApiKeys USED_API_KEY = ApiKeys.METADATA;
+    private static final Collection<ApiVersionsResponse.ApiVersion> EXPECTED_API_VERSIONS = createExpectedApiVersions();
+
+    private static Collection<ApiVersionsResponse.ApiVersion> createExpectedApiVersions() {
+        List<ApiVersionsResponse.ApiVersion> expectedApiVersions = new ArrayList<>();
+        expectedApiVersions.add(new ApiVersionsResponse.ApiVersion(USED_API_KEY.id, Protocol.MIN_VERSIONS[USED_API_KEY.id], Protocol.CURR_VERSION[USED_API_KEY.id]));
+        return expectedApiVersions;
+    }
+
+    @Override
+    protected Collection<ApiVersionsResponse.ApiVersion> getExpectedApiVersions() {
+        return EXPECTED_API_VERSIONS;
+    }
+
+    @Test
+    public void testUnsupportedLesserApiVersions() {
+        unsupportedApiVersionsCheck(getExpectedApiVersions(), (short) (ProtoUtils.latestVersion(USED_API_KEY.id) + 1), Short.MAX_VALUE, "Node 0 does not support required versions for Api " + USED_API_KEY.id);
+    }
+
+    @Test
+    public void testUnsupportedGreaterApiVersions() {
+        unsupportedApiVersionsCheck(getExpectedApiVersions(), Short.MIN_VALUE, (short) (ProtoUtils.oldestVersion(USED_API_KEY.id) - 1), "Node 0 does not support required versions for Api " + USED_API_KEY.id);
+    }
+
+    @Test
+    public void testUnsupportedMissingApiVersions() {
+        unsupportedApiVersionsCheck(new ArrayList<ApiVersionsResponse.ApiVersion>(), Short.MIN_VALUE, (short) (Short.MAX_VALUE + 2), "Node 0 does not support Api " + USED_API_KEY.id);
```

Comment:
I don't understand why we add +2 to `Short.MAX_VALUE` and then cast to `short` causing it to overflow?